pydantic/_internal/_typing_extra.py

```diff
@@ -1,192 +1,414 @@
-"""Logic for interacting with type annotations, mostly extensions, shims and hacks to wrap python's typing module."""
+"""Logic for interacting with type annotations, mostly extensions, shims and hacks to wrap Python's typing module."""
 
-from __future__ import annotations as _annotations
+from __future__ import annotations
 
-import dataclasses
+import collections.abc
 import re
 import sys
 import types
 import typing
 import warnings
-from collections.abc import Callable
-from functools import partial
-from types import GetSetDescriptorType
-from typing import TYPE_CHECKING, Any, Final
+from functools import lru_cache, partial
+from typing import Any, Callable
 
-from typing_extensions import Annotated, Literal, TypeAliasType, TypeGuard, Unpack, deprecated, get_args, get_origin
+import typing_extensions
+from typing_extensions import TypeIs, deprecated, get_args, get_origin
 
 from ._namespace_utils import GlobalsNamespace, MappingNamespace, NsResolver, get_module_ns_of
 
-if TYPE_CHECKING:
-    from ._dataclasses import StandardDataclass
+if sys.version_info < (3, 10):
+    NoneType = type(None)
+    EllipsisType = type(Ellipsis)
+else:
+    from types import EllipsisType as EllipsisType
+    from types import NoneType as NoneType
 
-try:
-    from typing import _TypingBase  # type: ignore[attr-defined]
-except ImportError:
-    from typing import _Final as _TypingBase  # type: ignore[attr-defined]
 
-typing_base = _TypingBase
+# See https://typing-extensions.readthedocs.io/en/latest/#runtime-use-of-types:
 
 
-if sys.version_info < (3, 9):
-    # python < 3.9 does not have GenericAlias (list[int], tuple[str, ...] and so on)
-    TypingGenericAlias = ()
-else:
-    from typing import GenericAlias as TypingGenericAlias  # type: ignore
+@lru_cache(maxsize=None)
+def _get_typing_objects_by_name_of(name: str) -> tuple[Any, ...]:
+    result = tuple(getattr(module, name) for module in (typing, typing_extensions) if hasattr(module, name))
+    if not result:
+        raise ValueError(f'Neither typing nor typing_extensions has an object called {name!r}')
+    return result
 
 
-if sys.version_info < (3, 11):
-    from typing_extensions import NotRequired, Required
-else:
-    from typing import NotRequired, Required  # noqa: F401
+# As suggested by the `typing-extensions` documentation, we could apply caching to this method,
+# but it doesn't seem to improve performance. This also requires `obj` to be hashable, which
+# might not be always the case:
+def _is_typing_name(obj: object, name: str) -> bool:
+    # Using `any()` is slower:
```

Comment:
Can we add a docstring to this function and the one above? If you're switching contexts into the world of typing and types management, a little description here could really help.